src/Kriti/CustomFunctions.hs

```diff
@@ -0,0 +1,82 @@
+module Kriti.CustomFunctions (basicFunctions) where
+
+import qualified Data.Aeson as J
+import qualified Data.Scientific as S
+import qualified Data.Text as T
+import qualified Data.Vector as V
+import Kriti.Error (CustomFunctionError (..))
+
+type KritiFunc = J.Value -> Either CustomFunctionError J.Value
+
+-- | Basic custom function for Kriti.
+basicFunctions :: [(T.Text, KritiFunc)]
+basicFunctions =
+  [ ("empty", emptyF),
+    ("size", lengthF),
+    ("inverse", inverseF),
+    ("head", headF),
+    ("tail", tailF),
+    ("toCaseFold", toCaseFoldF),
+    ("toLower", toLowerF),
+    ("toUpper", toUpperF),
+    ("toTitle", toTitleF)
+  ]
+
+emptyF :: KritiFunc
+emptyF inp = case inp of
+  J.Object km -> Right . J.Bool $ null km
+  J.Array vec -> Right . J.Bool $ null vec
+  J.String txt -> Right . J.Bool $ T.strip txt == ""
+  J.Number sci -> Right . J.Bool $ sci == 0
+  J.Bool _ -> Left . CustomFunctionError $ "Cannot define emptiness for a boolean"
+  J.Null -> Right $ J.Bool True
+
+lengthF :: KritiFunc
+lengthF inp = Right . J.Number $ case inp of
+  J.Object km -> (`S.scientific` 0) . toInteger $ length km
+  J.Array vec -> (`S.scientific` 0) . toInteger $ length vec
+  J.String txt -> (`S.scientific` 0) . toInteger $ T.length txt
+  J.Number sci -> sci
+  J.Bool b -> (`S.scientific` 0) . toInteger $ fromEnum b
+  J.Null -> 0
+
+inverseF :: KritiFunc
+inverseF inp = case inp of
+  J.Object km -> Right . J.Object $ km
+  J.Array vec -> Right . J.Array $ V.reverse vec
+  J.String txt -> Right . J.String $ T.reverse txt
+  J.Number sci -> Right . J.Number $ 1 / sci
+  J.Bool b -> Right . J.Bool $ not b
+  J.Null -> Right J.Null
+
+headF :: KritiFunc
+headF inp = case inp of
+  J.Array vec -> Right . V.head $ vec
+  J.String txt -> Right . J.String . T.singleton . T.head $ txt
```

Comment:
both of these functions should use the variants that return `Maybe` values so we can throw a kriti error rather than having it create an impure exception.

e.g. (a sketch, haven't type checked)
```haskell
case T.uncons txt of
  Just (a, _) -> Right . J.String . T.singleton $ a
  Nothing -> Left <err>
```

Reply:
Thanks Joe, I have updated the head function to throw kriti error.